README.rst
================================================================================
  MPC: Morphological Parser Creator
================================================================================

The MPC is the Morphological Parser Creator, a web service that can create
morphological parsers using finite-state transducers (FSTs) and N-gram language
models (LMs).


Creating a Morphophonology
================================================================================

JSON passed over HTTP::

    {"phonology-id": "28e4406c-a325-496e-967c-29e41f94a66f"
     "morphology-id": "49b04725-dcc0-45bc-b317-8e89896593bc"}

Clojure Map created by calling ``(parse-string json-string true)``::

    {:phonology-id: "28e4406c-a325-496e-967c-29e41f94a66f"
     :morphology-id: "49b04725-dcc0-45bc-b317-8e89896593bc"}



Basics
================================================================================

Serve the MPC locally::

    $ lein ring server-headless

Run the tests::

    $ lein test

Run the tests continuously and have them refresh as the code changes::

    $ lein test-refresh


Docker Stuff
================================================================================

This doesn't work currently.

Build the docker image::

    $ docker build -t mpc .

Run the docker image::

    $ docker run -it --rm --name my-running-app mpc


Database
================================================================================

MPC is using PostgreSQL as the database backend and Clojure's ragtime to manage
migrations.

To start the PostgreSQL server on a Mac::

    $ postgres -D /usr/local/var/postgres

To create the needed PostgreSQL database, make sure you have PostgreSQL
installed and then run::

    $ psql
    youruser=# CREATE USER mpc WITH PASSWORD 'mpc';
    youruser=# CREATE DATABASE mpc OWNER mpc;

To destroy the database::

    $ psql
    youruser=# DROP DATABASE mpc;

To use ragtime to apply the migrations::

    $ export MPC_DB_URI="jdbc:postgresql://localhost:5432/mpc?user=mpc&password=mpc"
    $ lein repl
    mpc.core=> (require '[mpc.migrate :refer :all])
    mpc.core=> (require '[ragtime.repl :as repl])
    (repl/migrate config)
    Applying 001-initial

To undo the migrations::

    mpc.core=> (repl/rollback config)


PostgreSQL Cheatsheet
--------------------------------------------------------------------------------

Open PostgreSQL command-line interface::

    $ psql
    youruser=#

Open PostgreSQL command-line interface and connect to ``mpc`` db::

    $ psql mpc
    mpc=#

Show databases::

    youruser=# \l
                                     List of databases
        Name    |   Owner    | Encoding |   Collate   |    Ctype    |     Access privileges
    ------------+------------+----------+-------------+-------------+---------------------------
     youruser   | youruser   | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
     mpc        | mpc        | UTF8     | en_US.UTF-8 | en_US.UTF-8 |

Connect to database ``mpc``::

    youruser=# \c mpc
    You are now connected to database "mpc" as user "youruser".
    mpc=#

Show tables::

    mpc=# \dt
                    List of relations
    Schema |         Name         | Type  |  Owner
    --------+----------------------+-------+----------
    public | candidate_ranker     | table | mpc
    public | morphological_parser | table | mpc
    public | morphology           | table | mpc
    public | morphophonology      | table | mpc
    public | phonology            | table | mpc
    public | ragtime_migrations   | table | mpc

Describe the ``phonology`` table::

    mpc=# \d+ phonology
                                                 Table "public.phonology"
         Column      |          Type          | Collation | Nullable | Default | Storage  | Stats target | Description 
    -----------------+------------------------+-----------+----------+---------+----------+--------------+-------------
     id              | integer                |           | not null |         | plain    |              | 
     script          | text                   |           |          |         | extended |              | 
     compiled_script | character varying(500) |           |          |         | extended |              | 
    Indexes:
        "phonology_pkey" PRIMARY KEY, btree (id)
    Referenced by:
        TABLE "morphophonology" CONSTRAINT "morphophonology_phonology_id_fkey" FOREIGN KEY (phonology_id) REFERENCES phonology(id)


Access Database via Clojure REPL and Korma
--------------------------------------------------------------------------------

.::

    $ lein repl
    mpc.core=> (require 'korma.db)
    mpc.core=> (korma.db/defdb db (korma.db/postgres {:db "mpc"
                                                      :user "mpc"
                                                      :password "mpc"}))


cURL client examples
================================================================================

Create a phonology::

    $ curl \
    $     -H "Content-Type: application/json" \
    $     -X POST \
    $     -d "{\"script\": \"define phonology ...\"}" \
    $     http://localhost:3000/phonologies

Get all phonologies::

    $ curl http://localhost:3000/phonologies | jq



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-07-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-07-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/mpc/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mpc/compare/0.1.0...0.1.1


project.clj
(defproject mpc "0.1.0-SNAPSHOT"
  :description "MPC is the Morphological Parser Creator, an application that
               allows users to build morphological parsers for natural
               languages using finite state transducer (FST) technology (foma)
               and N-gram language models (MITLM). Part of the OLD (Online
               Linguistic Database) project"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [cheshire "5.8.0"]
                 [compojure "1.6.0"]
                 [environ "1.0.0"]
                 [com.stuartsierra/component "0.3.2"]
                 [hiccup "1.0.5"]
                 [inflections "0.13.0"]
                 [korma "0.4.3"]
                 [log4j "1.2.15" :exclusions [javax.mail/mail
                                              javax.jms/jms
                                              com.sun.jdmk/jmxtools
                                              com.sun.jmx/jmxri]]
                 [org.postgresql/postgresql "9.4.1208"]
                 [ragtime "0.7.2"]
                 [ring "1.6.3"]
                 [ring/ring-defaults "0.3.1"]
                 [ring/ring-jetty-adapter "1.6.3"]
                 [ring/ring-mock "0.3.2"]
                 [ring/ring-json "0.4.0"]]
  :main ^:skip-aot mpc.core
  :target-path "target/%s"
  :profiles {:dev {:source-paths ["dev"]
                   :env {:http-port 3000}}
                   :dependencies [[javax.servlet/servlet-api "2.5"]
                                  [ring/ring-mock "0.3.2"]]
             :prod {:env {:http-port 8000
                          :repl-port 8001}
                    :dependencies [[org.clojure/tools.nrepl "0.2.12"]]}
             :test {:ragtime {:database
                              "jdbc:postgresql://localhost:5432/mpc?user=mpc&password=mpc"}}
             :uberjar {:aot :all}}
  :plugins [[lein-environ "1.0.0"]
            ; The lein-ring plugin allows us to easily start a development web
            ; server with "lein ring server". It also allows us to package up
            ; our application as a standalone .jar or as a .war for deployment
            ; to a servlet container.
            [lein-ring "0.12.2"]
            [com.jakemccrary/lein-test-refresh "0.22.0"]
            [ragtime/ragtime.lein "0.3.9"]]
  :ring {:handler mpc.app/app
         :nrepl {:start? true
                 :port 9998}}
  :ragtime {:migrations ragtime.sql.files/resources/migrations
            :database (System/getenv "MPC_DB_URI")}
)
